---
title: "T1D Application"
output: rmarkdown::html_vignette
vignette: >
  %\VignetteIndexEntry{cFDR T1D application}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
---

```{r, echo = FALSE}
embed_png <- function(path, dpi = NULL) {
  meta <- attr(png::readPNG(path, native = TRUE, info = TRUE), "info")
  if (!is.null(dpi)) meta$dpi <- rep(dpi, 2)
  knitr::asis_output(paste0(
    "<img src='", path, "'",
    " width=", round(meta$dim[1] / (meta$dpi[1] / 96)),
    " height=", round(meta$dim[2] / (meta$dpi[2] / 96)),
    " />"
  ))
}

knitr::opts_chunk$set(comment = "#>", collapse = TRUE)
```

In this vignette, we walk through an example to illustrate how the `fcfdr` R package can be used to leverage various functional genomic data with GWAS $p$ values for type 1 diabetes (T1D) to find new associations.

**This example uses real GWAS and functional genomic data to generate $v$-values for ~114,000 SNPs and can therefore take up to 2 hours to run.**

---

The data required for this example is available to download within the `fcfdr` R package and includes:

1. GWAS $p$-values for T1D (Onengut-Gumuscu et al. (2015), https://www.nature.com/articles/ng.3245) downloaded from the GWAS Catalog (https://www.ebi.ac.uk/gwas/publications/25751624)

2. GWAS $p$-values for Rheumatoid Arthritis (RA) (Eyre et al. (2012), https://www.nature.com/articles/ng.2462) downloaded from the GWAS Catalog (https://www.ebi.ac.uk/gwas/publications/23143596)

3. Binary measure of general chromatin accessibility at each SNP, derived from merging all DNaseI digital genomic footprinting (DGF) regions from the narrow-peak classifications across 57 cell types (https://www.nature.com/articles/nature11247; https://doi.org/10.1016/j.ajhg.2014.10.004). SNP annotations were downloaded for all 1000 Genomes phase 3 SNPs from the LDSC data repository (https://storage.googleapis.com/broad-alkesgroup-public/LDSCORE/1000G_Phase3_baselineLD_v2.1_ldscores.tgz) and the binary `DGF_ENCODE` annotation was extracted for all T1D SNPs in our analysis.

4. Fold-enrichment ratio of H3K27ac ChIP-seq counts relative to expected background counts in naive CD4+ T helper cells (https://www.nature.com/articles/nbt1010-1045). Downloaded from https://egg2.wustl.edu/roadmap/data/byFileType/signal/consolidated/macs2signal/foldChange/E043-H3K27ac.fc.signal.bigwig.

---

First, we download the data:

```{r eval = FALSE}
set.seed(1)
library(fcfdr)
data(df, package = "fcfdr")
head(df)
```

In this application we leverage GWAS $p$-values for RA (`q1`), binary chromatin accessibility data (`q2`) and H3K27ac counts in naive CD4+ T helper cells (`q3`) with GWAS $p$-values for T1D (`orig_p`) to generate $v$-values. 

```{r eval = FALSE}
orig_p <- df$p
chr <- df$chrom
q1 <- df$RA_p
q2 <- df$DGF_ENCODE
q3 <- log(df$Th_H3K27ac+1)
```

---

The data frame also contains a column of LDAK weights for each SNP (http://dougspeed.com/calculate-weightings/). An LDAK weight of zero means that the signal is (almost) perfectly captured by neighbouring SNPs and so we use the subset of SNPs with non-zero LDAK weights as our independent subset of SNPs.

```{r eval = FALSE}
ind_snps <- which(df$ldak_weight != 0)
```

---

We are now ready to use the `fcfdr` R package to generate $v$-values. Firstly, we generate $v$-values leveraging GWAS $p$-values for RA. 

```{r eval = FALSE}
iter1_res <- flexible_cfdr(p = orig_p, 
                           q = q1, 
                           indep_index = ind_snps)
v1 <- iter1_res[[1]]$v
```

---

The resultant $v$-values for this first iteration (`v1`) are then used in the next iteration to leverage binary chromatin accessibility data. Note that the binary cFDR function implements a leave-one-chromosome out procedure and therefore requires the chromosomes for each SNP rather than an independent subset.

```{r eval = FALSE}
iter2_res <- binary_cfdr(p = v1,
                         q = q2,
                         chr = chr)

v2 <- iter2_res$v
```

---

The resultant $v$-values for this second iteration (`v2`) are then used in the next iteration to leverage H3K27ac counts.

```{r eval = FALSE}
iter3_res <- flexible_cfdr(p = v2, 
                           q = q3, 
                           indep_index = ind_snps)

v3 <- iter3_res[[1]]$v
```

---

We then create a final data frame for the `fcfdr` results. Note that the sign is flipped for $q2$ and $q3$. This is because these are negatively correlated with `p` and the flexible cFDR software automatically flips the sign of `q` to ensure that low `p` are enriched for low `q`.

```{r eval = FALSE}
res <- data.frame(orig_p, q1 = iter1_res[[1]]$q, q2 = as.factor(iter2_res$q), q3 = iter3_res[[1]]$q, v1, v2, v3)

head(res)
```

---

We inspect the resultant $v$-values after each iteration by plotting these against the values used as `p` in that iteration.

```{r eval = FALSE}
library(ggplot2)
library(cowplot)

mid1 <- median(res$q1)

ggplot(res, aes(x = orig_p, y = v1, col = q1)) + geom_point(cex = 0.5) + theme_cowplot(12) + background_grid(major = "xy", minor = "none") + geom_abline(intercept = 0, slope = 1,  linetype="dashed") + xlab("P") + ylab("V1") + ggtitle(paste0("Iteration 1")) + scale_color_gradient2(midpoint = mid1, low = "blue", mid = "white", high = "red", space = "Lab")
```

```{r, echo = FALSE}
embed_png("iter1_plot.png")
```

```{r eval = FALSE}
ggplot(res, aes(x = v1, y = v2, col = q2)) + geom_point(cex = 0.5) + theme_cowplot(12) + background_grid(major = "xy", minor = "none") + geom_abline(intercept = 0, slope = 1,  linetype="dashed") + xlab("V1") + ylab("V2") + ggtitle(paste0("Iteration 2"))
```

```{r, echo = FALSE}
embed_png("iter2_plot.png")
```

```{r eval = FALSE}
mid3 <- median(res$q3)

ggplot(res, aes(x = v2, y = v3, col = q3)) + geom_point(cex = 0.5) + theme_cowplot(12) + background_grid(major = "xy", minor = "none") + geom_abline(intercept = 0, slope = 1,  linetype="dashed") + xlab("V2") + ylab("V3") + ggtitle(paste0("Iteration 3")) + scale_color_gradient2(midpoint = mid3, low = "blue", mid = "white", high = "red", space = "Lab")
```

```{r, echo = FALSE}
embed_png("iter3_plot.png")
```

---

We can also plot the original $p$-values for T1D against the final adjusted $v$-values.

```{r eval = FALSE}
mid1 <- median(res$q1)

ggplot(res, aes(x = orig_p, y = v3)) + geom_point(cex = 0.5, alpha = 0.5) + theme_cowplot(12) + background_grid(major = "xy", minor = "none") + geom_abline(intercept = 0, slope = 1,  linetype="dashed", col = "red") + xlab("P") + ylab("V3") + ggtitle(paste0("V3 against original p-values"))
```

```{r, echo = FALSE}
embed_png("v3_p.png")
```

```{r eval = FALSE}
ggplot(res, aes(x = -log10(orig_p), y = -log10(v3))) + geom_point(cex = 0.5, alpha = 0.5) + theme_cowplot(12) + background_grid(major = "xy", minor = "none") + geom_abline(intercept = 0, slope = 1,  linetype="dashed", col = "red") + xlab("P (-log10)") + ylab("V3 (-log10)") + ggtitle(paste0("V3 against original p-values (-log10)")) + coord_cartesian(ylim = c(0,10), xlim = c(0,10))
```

```{r, echo = FALSE}
embed_png("log10_v3_p.png")
```

---

Finally, we find that flexible cFDR identifies newly FDR significant SNPs that have relatively small GWAS $p$-values for rheumatoid arthritis, are more likely to be found in open chromatin regions and have relatively high H3K27ac counts in a T1D relevant cell type.

```{r eval = FALSE}
fdr_thr <- 4*10^-6

length(which(v3 < fdr_thr & orig_p > fdr_thr))

median(df$RA_p[which(v3 < fdr_thr & orig_p > fdr_thr)])
median(df$RA_p)

mean(df$DGF_ENCODE[which(v3 < fdr_thr & orig_p > fdr_thr)])
mean(df$DGF_ENCODE)

median(df$Th_H3K27ac[which(v3 < fdr_thr & orig_p > fdr_thr)])
median(df$Th_H3K27ac)
```
