<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>Extra Information • fcfdr</title>
<!-- favicons --><link rel="icon" type="image/png" sizes="16x16" href="../favicon-16x16.png">
<link rel="icon" type="image/png" sizes="32x32" href="../favicon-32x32.png">
<link rel="apple-touch-icon" type="image/png" sizes="180x180" href="../apple-touch-icon.png">
<link rel="apple-touch-icon" type="image/png" sizes="120x120" href="../apple-touch-icon-120x120.png">
<link rel="apple-touch-icon" type="image/png" sizes="76x76" href="../apple-touch-icon-76x76.png">
<link rel="apple-touch-icon" type="image/png" sizes="60x60" href="../apple-touch-icon-60x60.png">
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><meta property="og:title" content="Extra Information">
<meta property="og:description" content="fcfdr">
<meta property="og:image" content="/logo.png">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body data-spy="scroll" data-target="#toc">
    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">fcfdr</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">1.0.0</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../index.html">
    <span class="fas fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="../articles/extra-information.html">Extra Information</a>
    </li>
    <li>
      <a href="../articles/fcfdr-vignette.html">Introductory vignette</a>
    </li>
    <li>
      <a href="../articles/ldak-vignette.html">LDAK vignette</a>
    </li>
    <li>
      <a href="../articles/t1d_app.html">T1D Application</a>
    </li>
  </ul>
</li>
<li>
  <a href="../news/index.html">Changelog</a>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right"></ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><script src="extra-information_files/header-attrs-2.6/header-attrs.js"></script><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1 data-toc-skip>Extra Information</h1>
            
      
      
      <div class="hidden name"><code>extra-information.Rmd</code></div>

    </div>

    
    
<p>This vignette contains additional details for the <code>fcfdr</code> R package.</p>
<div id="contents" class="section level3">
<h3 class="hasAnchor">
<a href="#contents" class="anchor"></a>Contents</h3>
<ol style="list-style-type: decimal">
<li><a href="#GC">Genomic control</a></li>
<li><a href="#AD">Auxiliary data</a></li>
<li><a href="#ind">Finding independent SNPs</a></li>
<li><a href="#iter">Applying cFDR iteratively</a></li>
</ol>
<hr>
</div>
<div id="GC" class="section level3">
<h3 class="hasAnchor">
<a href="#GC" class="anchor"></a>Genomic control</h3>
<p>A key assumption in the cFDR methodology is that the <span class="math inline">\(p\)</span>-values are uniformly distributed under the null hypothesis of no association. Therefore, prior to applying cFDR, users should check that this assumption is satisfied by calculating the genomic inflation factor, <span class="math inline">\(\lambda\)</span>, and applying genomic control if necessary.</p>
<hr>
</div>
<div id="AD" class="section level3">
<h3 class="hasAnchor">
<a href="#AD" class="anchor"></a>Auxiliary data</h3>
<p>The distinguishing feature of <code>fcfdr</code> compared to earlier cFDR methods is that it can leverage auxiliary data from any arbitrary distribution. Some examples of auxiliary covariates to leverage include:</p>
<ul>
<li><p><strong>Functional genomic assays.</strong> It is known that GWAS SNPs are not randomly distributed across functional categories, for example GWAS SNPs are typically enriched in enhancer and open chromatin regions in relevant cell types. This suggests that leveraging (e.g.) ATAC-seq data or ChIP-seq data for histone modifications (e.g. H3K27ac marking enhancer regions) in relevant cell types may be useful. If the relevant cell types are known then we found the <a href="https://egg2.wustl.edu/roadmap/data/byFileType/signal/consolidated/macs2signal/foldChange/">consolidated fold change values from NIH Roadmap</a> to be useful (epigenome ID to cell type conversion available <a href="https://docs.google.com/spreadsheets/d/1yikGx4MsO9Ei36b64yOy9Vb6oPC5IBGlFbYEt-N6gOM/edit#gid=15">here</a>). GWAS SNPs can easily be matched to the fold change value using the bedtools intersect function with the <code>-wb</code> tag.</p></li>
<li><p><strong>Per-SNP scores of pathogenicity/ functionality/ deleteriousness.</strong> Many tools have been developed that integrate various genomic and epigenomic annotation data to quantify the pathogenicity, functionality and/or deleteriousness of both coding and non-coding GWAS variants. For example tissue-specific <a href="https://pubmed.ncbi.nlm.nih.gov/28742084/%5D">GenoSkyline scores</a> or <a href="https://genomebiology.biomedcentral.com/articles/10.1186/s13059-018-1546-6">PINES scoring</a>.</p></li>
<li><p><strong>LD score annotations</strong> The latest and recommended baseline-LD model v2.2 contains 97 annotations ranging from binary synonymous/non-synonymous annotations to continuous functional genomic annotations. The annotations for all 1000 genomes phase 3 SNPs (note: now extended to 19,476,620 UK Biobank SNPs with <span class="math inline">\(MAF\geq 0.1%\)</span>) can be downloaded readily online.</p></li>
</ul>
<hr>
</div>
<div id="ind" class="section level3">
<h3 class="hasAnchor">
<a href="#ind" class="anchor"></a>Finding independent subsets of SNPs</h3>
<p>The <code><a href="../reference/flexible_cfdr.html">flexible_cfdr()</a></code> function requires the indices of an independent subset of the GWAS SNPs to ensure non-biased bandwidth estimation for the KDE. In practice, we use SNPs allocated a non-zero LDAK weight as our subset of independent SNPs. Instructions to generate LDAK weights can be found <a href="https://annahutch.github.io/fcfdr/articles/ldak-vignette.html">in this vignette</a>. Alternatively, PLINK can also be used (see <a href="https://www.cog-genomics.org/plink/1.9/ld">here</a>).</p>
<p>The binary cFDR method does not fit a KDE and instead uses empirical CDFs. For this reason, a leave-one-chromosome-out procedure is implemented intrinsically in the <code><a href="../reference/binary_cfdr.html">binary_cfdr()</a></code> function. Therefore, the <code><a href="../reference/binary_cfdr.html">binary_cfdr()</a></code> function requires per-SNP chromosomes as input.</p>
<hr>
</div>
<div id="iter" class="section level3">
<h3 class="hasAnchor">
<a href="#iter" class="anchor"></a>Applying cFDR iteratively</h3>
<p>Our cFDR method can be applied iteratively, whereby the <span class="math inline">\(v\)</span>-values from the previous iteration are used as the <span class="math inline">\(p\)</span>-values in the next iteration. This allows additional layers of information to be incorporated into the analysis.</p>
<p>An example of this is shown in the <a href="https://annahutch.github.io/fcfdr/articles/t1d_app.html">type 1 diabetes vignette</a>.</p>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">

      </div>

</div>



      <footer><div class="copyright">
  <p>Developed by Anna Hutchinson, Chris Wallace, Thomas Willis.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="https://pkgdown.r-lib.org/">pkgdown</a> 1.6.1.</p>
</div>

      </footer>
</div>

  


  </body>
</html>
