<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>Introductory vignette • fcfdr</title>
<!-- favicons --><link rel="icon" type="image/png" sizes="16x16" href="../favicon-16x16.png">
<link rel="icon" type="image/png" sizes="32x32" href="../favicon-32x32.png">
<link rel="apple-touch-icon" type="image/png" sizes="180x180" href="../apple-touch-icon.png">
<link rel="apple-touch-icon" type="image/png" sizes="120x120" href="../apple-touch-icon-120x120.png">
<link rel="apple-touch-icon" type="image/png" sizes="76x76" href="../apple-touch-icon-76x76.png">
<link rel="apple-touch-icon" type="image/png" sizes="60x60" href="../apple-touch-icon-60x60.png">
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><meta property="og:title" content="Introductory vignette">
<meta property="og:description" content="fcfdr">
<meta property="og:image" content="/logo.png">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body data-spy="scroll" data-target="#toc">
    <div class="container template-article">
      <header><div class="navbar navbar-inverse navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">fcfdr</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">0.0.1</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../index.html">
    <span class="fas fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="../articles/fcfdr-vignette.html">Introductory vignette</a>
    </li>
    <li>
      <a href="../articles/ldak-vignette.html">Generating independent SNP set</a>
    </li>
    <li>
      <a href="../articles/extra-information.html">Extra information</a>
    </li>
  </ul>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right">
<li>
  <a href="https://github.com/annahutch/fcfdr">
    <span class=" github"></span>
     
  </a>
</li>
      </ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><script src="fcfdr-vignette_files/accessible-code-block-0.0.1/empty-anchor.js"></script><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1 data-toc-skip>Introductory vignette</h1>
            
      
      
      <div class="hidden name"><code>fcfdr-vignette.Rmd</code></div>

    </div>

    
    
<p>The <code>fcfdr</code> R package implements cFDR for a variety of auxiliary covariates, unlike earlier empirical cFDR methods (Liley and Wallace 2019; <a href="https://github.com/jamesliley/cfdr" class="uri">https://github.com/jamesliley/cfdr</a>) that only support auxiliary <span class="math inline">\(p\)</span>-values from related traits. A direct utility of <code>fcfdr</code> is to leverage relevant functional genomic data with GWAS <span class="math inline">\(p\)</span>-values to increase power for GWAS discovery. The method generates “v-values” which can be interpreted as GWAS <span class="math inline">\(p\)</span>-values that have been re-weighted according to the auxiliary data values. Since the v-values are analogous to p-values, they can be used directly in any error-rate controlling procedure.</p>
<p>The <code>fcfdr</code> R package contains two key functions:</p>
<ol style="list-style-type: decimal">
<li><p><a href="https://annahutch.github.io/fcfdr/reference/flexible_cfdr.html"><code>flexible_cfdr()</code></a>: Implements cFDR leveraging <span class="math inline">\(\textbf{continuous}\)</span> auxiliary covariates.</p></li>
<li><p><a href="https://annahutch.github.io/fcfdr/reference/binary_cfdr.html"><code>binary_cfdr()</code></a>: Implements cFDR leveraging <span class="math inline">\(\textbf{binary}\)</span> auxiliary covariates.</p></li>
</ol>
<hr>
<p>Both functions require 2 parameters to be specified:</p>
<ul>
<li><p><code>p</code>: GWAS <span class="math inline">\(p\)</span>-values for the trait of interest (vector of per-SNP <span class="math inline">\(p\)</span>-values)</p></li>
<li><p><code>q</code>: Auxiliary data values (vector of per-SNP auxiliary data values)</p></li>
</ul>
<p>along with either the indices of a subset of independent SNPs (<code>indep_index</code>) in <code><a href="../reference/flexible_cfdr.html">flexible_cfdr()</a></code> or the chromosome for which each SNP resides (<code>chr</code>) in <code><a href="../reference/binary_cfdr.html">binary_cfdr()</a></code>.</p>
<p>For further details, including instructions to generate <code>indep_index</code>, examples of auxiliary data to leverage and instructions to apply cFDR iteratively, please see the <a href="https://annahutch.github.io/fcfdr/articles/extra-information.html">Extra Information vignette</a></p>
<hr>
<div id="simple-walk-through-example" class="section level4">
<h4 class="hasAnchor">
<a href="#simple-walk-through-example" class="anchor"></a>Simple walk-through example</h4>
<p>Firstly, load the flexible cFDR package:</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/r/base/Random.html">set.seed</a></span><span class="op">(</span><span class="fl">1</span><span class="op">)</span>
<span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va">fcfdr</span><span class="op">)</span></code></pre></div>
<hr>
<p>Next, simulate <span class="math inline">\(p\)</span>-values for 50,000 genetic variants, including 500 associated variants.</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">n</span> <span class="op">=</span> <span class="fl">50000</span>
<span class="va">n1p</span> <span class="op">=</span> <span class="fl">500</span> <span class="co"># associated variants</span>
<span class="va">zp</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/stats/Normal.html">rnorm</a></span><span class="op">(</span><span class="va">n1p</span>, sd<span class="op">=</span><span class="fl">5</span><span class="op">)</span>, <span class="fu"><a href="https://rdrr.io/r/stats/Normal.html">rnorm</a></span><span class="op">(</span><span class="va">n</span><span class="op">-</span><span class="va">n1p</span>, sd<span class="op">=</span><span class="fl">1</span><span class="op">)</span><span class="op">)</span> <span class="co"># z-scores</span>
<span class="va">p</span> <span class="op">=</span> <span class="fl">2</span><span class="op">*</span><span class="fu"><a href="https://rdrr.io/r/stats/Normal.html">pnorm</a></span><span class="op">(</span><span class="op">-</span><span class="fu"><a href="https://rdrr.io/r/base/MathFun.html">abs</a></span><span class="op">(</span><span class="va">zp</span><span class="op">)</span><span class="op">)</span> <span class="co"># convert to p-values</span>
<span class="fu"><a href="https://rdrr.io/r/graphics/hist.html">hist</a></span><span class="op">(</span><span class="va">p</span><span class="op">)</span></code></pre></div>
<p><img src="fcfdr-vignette_files/figure-html/unnamed-chunk-3-1.png" width="700"></p>
<hr>
<p>We simulate relevant auxiliary data from a mixture normal distribution (<span class="math inline">\(q\)</span>). The associated SNPs (with indices 1-500) are sampled from <span class="math inline">\(N(-0.5, 0.5^2)\)</span> and the non-associated SNPs (with indices 500-50000) are sampled from <span class="math inline">\(N(2,1)\)</span>.</p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">mixture_comp1</span> <span class="op">&lt;-</span> <span class="kw">function</span><span class="op">(</span><span class="va">x</span><span class="op">)</span> <span class="fu"><a href="https://rdrr.io/r/stats/Normal.html">rnorm</a></span><span class="op">(</span><span class="va">x</span>, mean <span class="op">=</span> <span class="op">-</span><span class="fl">0.5</span>, sd <span class="op">=</span> <span class="fl">0.5</span><span class="op">)</span>
<span class="va">mixture_comp2</span> <span class="op">&lt;-</span> <span class="kw">function</span><span class="op">(</span><span class="va">x</span><span class="op">)</span> <span class="fu"><a href="https://rdrr.io/r/stats/Normal.html">rnorm</a></span><span class="op">(</span><span class="va">x</span>, mean <span class="op">=</span> <span class="fl">2</span>, sd <span class="op">=</span> <span class="fl">1</span><span class="op">)</span>
<span class="va">n</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/length.html">length</a></span><span class="op">(</span><span class="va">p</span><span class="op">)</span>
<span class="va">z</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/stats/Uniform.html">runif</a></span><span class="op">(</span><span class="va">n</span><span class="op">)</span>

<span class="va">q</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="fu">mixture_comp1</span><span class="op">(</span><span class="va">n1p</span><span class="op">)</span>, <span class="fu">mixture_comp2</span><span class="op">(</span><span class="va">n</span><span class="op">-</span><span class="va">n1p</span><span class="op">)</span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/r/graphics/hist.html">hist</a></span><span class="op">(</span><span class="va">q</span><span class="op">)</span></code></pre></div>
<p><img src="fcfdr-vignette_files/figure-html/unnamed-chunk-4-1.png" width="700"></p>
<hr>
<p>We can use the <code>corr_plot</code> function to visualise the relationship between <span class="math inline">\(p\)</span> and <span class="math inline">\(q\)</span>. We observe that low <span class="math inline">\(p\)</span>-values (i.e. high <span class="math inline">\(-log10(p)\)</span>) are enriched for low <span class="math inline">\(q\)</span> values.</p>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="../reference/corr_plot.html">corr_plot</a></span><span class="op">(</span><span class="va">p</span>, <span class="va">q</span><span class="op">)</span></code></pre></div>
<p><img src="fcfdr-vignette_files/figure-html/unnamed-chunk-5-1.png" width="576"></p>
<hr>
<p>We are now ready to use the <code>flexible_cfdr</code> function to derive the <span class="math inline">\(v\)</span>-values. Note that for the purpose of the vignette, we do not specify an independent subset of SNPs, however for real analyses this parameter should be specified appropriately to avoid biased bandwidth estimations when fitting the KDE. Subsets of independent SNPs can be readily found using PLINK or LDAK - see the <a href="http://dougspeed.com/calculate-weightings/">vignette for deriving LDAK weights here</a>.</p>
<p>This line of code will take approximately 4 minutes to run.</p>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">res</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/flexible_cfdr.html">flexible_cfdr</a></span><span class="op">(</span><span class="va">p</span>, <span class="va">q</span>, indep_index <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/seq.html">seq</a></span><span class="op">(</span><span class="fl">1</span>, <span class="va">n</span>, <span class="fl">1</span><span class="op">)</span><span class="op">)</span></code></pre></div>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/r/utils/str.html">str</a></span><span class="op">(</span><span class="va">res</span><span class="op">)</span>
<span class="co">#&gt; List of 2</span>
<span class="co">#&gt;  $ :'data.frame':    50000 obs. of  3 variables:</span>
<span class="co">#&gt;   ..$ p: num [1:50000] 1.73e-03 3.59e-01 2.94e-05 1.51e-15 9.94e-02 ...</span>
<span class="co">#&gt;   ..$ q: num [1:50000] -0.325 -0.297 -0.458 -0.383 -0.578 ...</span>
<span class="co">#&gt;   ..$ v: num [1:50000] 2.07e-04 4.94e-02 3.48e-06 1.70e-16 1.22e-02 ...</span>
<span class="co">#&gt;  $ :'data.frame':    1 obs. of  3 variables:</span>
<span class="co">#&gt;   ..$ q_low     : num -0.169</span>
<span class="co">#&gt;   ..$ left_cens : int 1108</span>
<span class="co">#&gt;   ..$ splinecorr: int 109</span>

<span class="va">p</span> <span class="op">=</span> <span class="va">res</span><span class="op">[[</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span><span class="op">$</span><span class="va">p</span>
<span class="va">q</span> <span class="op">=</span> <span class="va">res</span><span class="op">[[</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span><span class="op">$</span><span class="va">q</span>
<span class="va">v</span> <span class="op">=</span> <span class="va">res</span><span class="op">[[</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span><span class="op">$</span><span class="va">v</span></code></pre></div>
<p>The output from the function is a list of length two. The first element is a data.frame containing the <span class="math inline">\(p\)</span>-values (input parameter <code>p</code>), the auxiliary data values (input parameter <code>q</code>) and the generated <span class="math inline">\(v\)</span>-values. The second element contains auxiliary data, such as how many data-points were left-censored in the method and/or spline corrected.</p>
<p><strong>Note that the cFDR framework requires that low <span class="math inline">\(p\)</span>-values are enriched for low <span class="math inline">\(q\)</span> values, so that if the correlation between <span class="math inline">\(p\)</span> and <span class="math inline">\(q\)</span> is negative then the function intrinsically flips the sign of <span class="math inline">\(q\)</span>, meaning that the <span class="math inline">\(q\)</span> values reported in the data.frame output may be <span class="math inline">\(q:=-q\)</span>.</strong></p>
<hr>
<p>We can then visualise the results using the <code>pv_plot</code> and <code>log10pv_plot</code> functions. We observe that <span class="math inline">\(v\)</span>-values for SNPs with high <span class="math inline">\(q\)</span> values are increased, whilst those for SNPs with low <span class="math inline">\(q\)</span> values are decreased.</p>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="../reference/pv_plot.html">pv_plot</a></span><span class="op">(</span>p <span class="op">=</span> <span class="va">p</span>, q <span class="op">=</span> <span class="va">q</span>, v <span class="op">=</span> <span class="va">v</span><span class="op">)</span></code></pre></div>
<p><img src="fcfdr-vignette_files/figure-html/unnamed-chunk-8-1.png" width="700"></p>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="../reference/log10pv_plot.html">log10pv_plot</a></span><span class="op">(</span>p <span class="op">=</span> <span class="va">p</span>, q <span class="op">=</span> <span class="va">q</span>, v <span class="op">=</span> <span class="va">v</span>, 
             axis_lim <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="fl">0</span>, <span class="fl">10</span><span class="op">)</span><span class="op">)</span> <span class="co"># zoom in to interesting region</span></code></pre></div>
<p><img src="fcfdr-vignette_files/figure-html/unnamed-chunk-8-2.png" width="700"></p>
<hr>
<p>Finally, we run the Benjamini-Hochberg procedure on the <span class="math inline">\(v\)</span>-values and control the FDR at 0.05. This means that we are willing to accept up to 5% of the associations to be false positives.</p>
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">hit</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/which.html">which</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/stats/p.adjust.html">p.adjust</a></span><span class="op">(</span><span class="va">v</span>, method <span class="op">=</span> <span class="st">"BH"</span><span class="op">)</span> <span class="op">&lt;=</span> <span class="fl">0.05</span><span class="op">)</span></code></pre></div>
<p>For comparison, we do the same to the raw p-values:</p>
<div class="sourceCode" id="cb10"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">hit_p</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/which.html">which</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/stats/p.adjust.html">p.adjust</a></span><span class="op">(</span><span class="va">p</span>, method <span class="op">=</span> <span class="st">"BH"</span><span class="op">)</span> <span class="op">&lt;=</span> <span class="fl">0.05</span><span class="op">)</span></code></pre></div>
<p>‘True’ associations are those with indices 1-500, so the proportions of false discoveries are</p>
<div class="sourceCode" id="cb11"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># cFDR</span>
<span class="fl">1</span> <span class="op">-</span> <span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/length.html">length</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/sets.html">intersect</a></span><span class="op">(</span><span class="va">hit</span>,<span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="fl">1</span><span class="op">:</span><span class="fl">500</span><span class="op">)</span><span class="op">)</span><span class="op">)</span><span class="op">/</span><span class="fu"><a href="https://rdrr.io/r/base/length.html">length</a></span><span class="op">(</span><span class="va">hit</span><span class="op">)</span><span class="op">)</span> 
<span class="co">#&gt; [1] 0.04363636</span>

<span class="co"># p-value</span>
<span class="fl">1</span> <span class="op">-</span> <span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/length.html">length</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/sets.html">intersect</a></span><span class="op">(</span><span class="va">hit_p</span>,<span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="fl">1</span><span class="op">:</span><span class="fl">500</span><span class="op">)</span><span class="op">)</span><span class="op">)</span><span class="op">/</span><span class="fu"><a href="https://rdrr.io/r/base/length.html">length</a></span><span class="op">(</span><span class="va">hit_p</span><span class="op">)</span><span class="op">)</span>
<span class="co">#&gt; [1] 0.02262443</span></code></pre></div>
<hr>
<p>Altogether, the cFDR method has found 47 new associations that are true whilst controlling the FDR.</p>
<div class="sourceCode" id="cb12"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># number of extra true associations identified by flexible cFDR</span>
<span class="fu"><a href="https://rdrr.io/r/base/length.html">length</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/which.html">which</a></span><span class="op">(</span><span class="va">hit</span><span class="op">[</span><span class="op">!</span><span class="va">hit</span> <span class="op">%in%</span> <span class="va">hit_p</span><span class="op">]</span> <span class="op">&lt;=</span> <span class="fl">500</span><span class="op">)</span><span class="op">)</span>
<span class="co">#&gt; [1] 47</span></code></pre></div>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">

      </div>

</div>



      <footer><div class="copyright">
  <p>Developed by Anna Hutchinson, Chris Wallace, Thomas Willis.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="https://pkgdown.r-lib.org/">pkgdown</a> 1.6.1.</p>
</div>

      </footer>
</div>

  


  </body>
</html>
